File: exporter.py
import io
import datetime
import xlsxwriter
import holidays
from logic import Project, Task

def generate_excel(project: Project, language: str = "EN") -> io.BytesIO:
    output = io.BytesIO()
    workbook = xlsxwriter.Workbook(output, {'in_memory': True})
    
    sheet_name = "Timeline" if language == "EN" else "Cronograma"
    sheet = workbook.add_worksheet(sheet_name)
    sheet.hide_gridlines(2) 

    # --- FORMATS ---
    fmt_header_base = {
        'bold': True,
        'font_color': 'white',
        'bg_color': '#203764',
        'border': 1,
        'align': 'center',
        'valign': 'vcenter',
        'font_name': 'Arial',
        'font_size': 10,
        'text_wrap': True
    }
    fmt_header = workbook.add_format(fmt_header_base)
    
    # Header for Holiday Weeks
    fmt_header_holiday = workbook.add_format(dict(fmt_header_base, bg_color='#5B9BD5')) # Lighter blue to indicate something special

    fmt_phase = workbook.add_format({
        'bold': True,
        'font_color': '#203764',
        'bg_color': '#E7E6E6',
        'border': 1,
        'align': 'left',
        'valign': 'vcenter',
        'font_name': 'Arial',
        'font_size': 10
    })

    fmt_cell = workbook.add_format({
        'border': 1,
        'align': 'center',
        'valign': 'vcenter',
        'font_name': 'Arial',
        'font_size': 10
    })

    fmt_date = workbook.add_format({
        'num_format': 'dd-mmm-yy',
        'border': 1,
        'align': 'center',
        'valign': 'vcenter',
        'font_name': 'Arial',
        'font_size': 10
    })

    fmt_bar = workbook.add_format({
        'bg_color': '#203764', 
        'border': 1,
        'align': 'center',
        'valign': 'vcenter'
    })
    
    # Milestone / Markers
    # Font size larger for symbols
    fmt_marker = workbook.add_format({
        'align': 'center',
        'valign': 'vcenter',
        'font_color': '#C00000', # Red
        'bold': True,
        'border': 1,
        'font_name': 'Segoe UI Symbol', # Good for unicode
        'font_size': 12
    })
    
    # Holiday Cell logic? Maybe shade vertical column?
    # We can't shade the whole column easily without defining a format for EVERY cell type relative to holidays.
    # Instead, we will mark the header and maybe put a small marker in empty cells? 
    # Or define a separate background for holiday weeks?
    # Let's use a light pattern for holiday weeks if empty?
    # For now, just Header distinction is safest to not break bar colors.

    # --- TRANSLATIONS ---
    headers = {
        "EN": ["ID", "Phase", "Task", "Duration (Wks)", "Start", "End"],
        "PT": ["ID", "Fase", "Tarefa", "Duração (Sem)", "Início", "Fim"]
    }
    week_prefix = "w/c" if language == "EN" else "Sem de"

    # --- DATE RANGE ---
    if not project.tasks:
        start_date = project.start_date
        end_date = project.start_date + datetime.timedelta(weeks=4)
        week_start_list = []
        current = start_date - datetime.timedelta(days=start_date.weekday())
        while current < end_date:
            week_start_list.append(current)
            current += datetime.timedelta(weeks=1)
    else:
        p_start = project.start_date
        ends = [t.computed_end for t in project.tasks if t.computed_end]
        if ends:
            p_end = max(ends)
        else:
            p_end = p_start + datetime.timedelta(weeks=4)
            
        current = p_start - datetime.timedelta(days=p_start.weekday())
        week_start_list = []
        # Add buffer
        while current < p_end + datetime.timedelta(weeks=1):
            week_start_list.append(current)
            current += datetime.timedelta(weeks=1)

    if not week_start_list:
        week_start_list = [project.start_date - datetime.timedelta(days=project.start_date.weekday())]

    # --- HOLIDAY DETECTION ---
    country_holidays = holidays.country_holidays(project.country)
    
    # Check each week for holidays
    week_has_holiday = []
    for ws in week_start_list:
        # Check Mon-Fri
        found = False
        for d in range(5):
            day = ws + datetime.timedelta(days=d)
            if day in country_holidays:
                found = True
                break
        week_has_holiday.append(found)

    # --- WRITE HEADER ---
    col_static = headers[language]
    for i, h in enumerate(col_static):
        sheet.write(0, i, h, fmt_header)
    
    for i, ws in enumerate(week_start_list):
        col_idx = len(col_static) + i
        label = f"{week_prefix}\n{ws.strftime('%d-%b')}"
        
        # Use holiday header if holiday in week
        fmt = fmt_header_holiday if week_has_holiday[i] else fmt_header
        
        sheet.write(0, col_idx, label, fmt)
        sheet.set_column(col_idx, col_idx, 5)

    sheet.set_column(0, 0, 8)  
    sheet.set_column(1, 1, 25) 
    sheet.set_column(2, 2, 50) 
    sheet.set_column(3, 3, 12) 
    sheet.set_column(4, 5, 12) 
    sheet.freeze_panes(1, 6) # Freeze header and static cols

    # --- WRITE TASKS ---
    row = 1
    
    for task in project.tasks:
        if not task.computed_start or not task.computed_end:
            continue
            
        sheet.write(row, 0, task.id, fmt_cell)
        sheet.write(row, 1, task.phase, fmt_phase) 
        sheet.write(row, 2, task.name, fmt_cell)
        sheet.write(row, 3, task.duration_weeks, fmt_cell)
        sheet.write(row, 4, task.computed_start, fmt_date)
        sheet.write(row, 5, task.computed_end, fmt_date)
        
        for i, ws in enumerate(week_start_list):
            col_idx = len(col_static) + i
            
            # Simple overlap logic for weekly view
            # If task spans this week's start
            
            start_active = task.computed_start <= ws
            # Strict strictly less than end date?
            # If end date is Monday, it doesn't span that week.
            end_active = ws < task.computed_end
            
            is_active = start_active and end_active
            
            cell_val = ""
            cell_fmt = fmt_cell # Default blank
            
            if is_active:
                # MARKER LOGIC
                # Only show marker in the LAST week of the task
                # Last week starts at (task.computed_end - 7 days)? No.
                # Task end date logic:
                # If Task ends Oct 25 (Friday). Week starts Oct 21. 
                # ws = Oct 21. computed_end = Oct 25 (or Oct 28 Monday).
                # If we assume computed_end is the start of the NEXT block (exclusive),
                # Then the week containing the task end is the one starting before it.
                
                # Check if this is the "concluding week"
                # The week that contains (computed_end - 1 day)
                end_minus_one = task.computed_end - datetime.timedelta(days=1)
                is_concluding_week = (ws <= end_minus_one < ws + datetime.timedelta(days=7))
                
                if is_concluding_week:
                     if task.task_type == "Milestone":
                         cell_val = "▲"
                         cell_fmt = fmt_marker
                     elif task.task_type == "Key Decision":
                         cell_val = "★"
                         cell_fmt = fmt_marker
                     elif task.task_type == "Bottleneck":
                         cell_val = "⚠️"
                         cell_fmt = fmt_marker
                     elif task.task_type == "External Dependency":
                         cell_val = "🔗"
                         cell_fmt = fmt_marker
                     else:
                         cell_fmt = fmt_bar
                else:
                     # Active but not concluding -> Fill for Standard/Bottleneck
                     # Milestones usually don't have duration bars unless specified.
                     if task.task_type not in ["Milestone", "Key Decision", "External Dependency"]:
                         cell_fmt = fmt_bar
            
            sheet.write(row, col_idx, cell_val, cell_fmt)
            
        row += 1

    workbook.close()
    output.seek(0)
    return output

File: src/lib/exporter.ts
import ExcelJS from 'exceljs';
import type { Project } from './logic';
import { addDays } from 'date-fns';
import Holidays from 'date-holidays';
import { saveAs } from 'file-saver';

export async function exportToExcel(project: Project, language: "EN" | "PT" = "EN") {
    try {
        const workbook = new ExcelJS.Workbook();
        const sheetName = language === "EN" ? "Timeline" : "Cronograma";
        const worksheet = workbook.addWorksheet(sheetName);

        // Styles
        const headerStyle: Partial<ExcelJS.Style> = {
            font: { bold: true, color: { argb: 'FFFFFFFF' }, name: 'Arial', size: 10 },
            fill: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF203764' } }, // Dark Blue
            alignment: { vertical: 'middle', horizontal: 'center', wrapText: true },
            border: { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } }
        };

        const holidayHeaderStyle: Partial<ExcelJS.Style> = {
            ...headerStyle,
            fill: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF5B9BD5' } } // Light Blue
        };

        const phaseStyle: Partial<ExcelJS.Style> = {
            font: { bold: true, color: { argb: 'FF203764' }, name: 'Arial', size: 10 },
            fill: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFE7E6E6' } }, // Light Grey
            alignment: { vertical: 'middle', horizontal: 'left' },
            border: { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } }
        };

        const cellStyle: Partial<ExcelJS.Style> = {
            font: { name: 'Arial', size: 10 },
            alignment: { vertical: 'middle', horizontal: 'center' },
            border: { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } }
        };

        const dateStyle: Partial<ExcelJS.Style> = {
            ...cellStyle,
            numFmt: 'dd-mmm-yy'
        };

        const barStyle: Partial<ExcelJS.Style> = {
            fill: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF203764' } },
            border: { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } }
        };

        const compressedBarStyle: Partial<ExcelJS.Style> = {
            fill: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFEF4444' } }, // Red
            border: { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } }
        };

        const markerStyle: Partial<ExcelJS.Style> = {
            font: { bold: true, color: { argb: 'FFC00000' }, name: 'Segoe UI Symbol', size: 12 },
            alignment: { vertical: 'middle', horizontal: 'center' },
            border: { top: { style: 'thin' }, left: { style: 'thin' }, bottom: { style: 'thin' }, right: { style: 'thin' } }
        };

        // Translations
        const headers = language === "EN"
            ? ["ID", "Phase", "Task", "Duration (Wks)", "Start", "End"]
            : ["ID", "Fase", "Tarefa", "Duração (Sem)", "Início", "Fim"];
        const weekPrefix = language === "EN" ? "w/c" : "Sem de";

        // Date Range
        let startDate = new Date(project.startDate);
        const day = startDate.getDay();
        const diff = startDate.getDate() - day + (day === 0 ? -6 : 1);
        let currentMonday = new Date(startDate.setDate(diff));
        currentMonday = new Date(project.startDate);
        const d = currentMonday.getDay();
        const diff2 = currentMonday.getDate() - d + (d === 0 ? -6 : 1);
        currentMonday = new Date(currentMonday.setDate(diff2));

        let maxEnd = new Date(currentMonday);
        maxEnd = addDays(maxEnd, 28);

        project.tasks.forEach(t => {
            if (t.computedEnd && new Date(t.computedEnd) > maxEnd) {
                maxEnd = new Date(t.computedEnd);
            }
        });

        const weeks: Date[] = [];
        let w = new Date(currentMonday);
        while (w < addDays(maxEnd, 7)) {
            weeks.push(new Date(w));
            w = addDays(w, 7);
        }

        // Holiday Detection (Multi-Calendar)
        const hds = project.country.map(c => {
            try { return new Holidays(c); } catch { return null; }
        }).filter(Boolean);

        const weekHasHoliday = weeks.map(ws => {
            for (let i = 0; i < 5; i++) {
                const d = addDays(ws, i);
                for (const hd of hds) {
                    if (hd && hd.isHoliday(d)) return true;
                }
            }
            return false;
        });

        // Write Headers
        headers.forEach((h, i) => {
            const cell = worksheet.getCell(1, i + 1);
            cell.value = h;
            cell.style = headerStyle;
        });

        weeks.forEach((ws, i) => {
            const colIdx = headers.length + i + 1;
            const cell = worksheet.getCell(1, colIdx);
            const dayStr = ws.getDate().toString().padStart(2, '0');
            const monthStr = ws.toLocaleString(language === "EN" ? 'en-US' : 'pt-PT', { month: 'short' });
            cell.value = `${weekPrefix}\n${dayStr}-${monthStr}`;
            cell.style = weekHasHoliday[i] ? holidayHeaderStyle : headerStyle;
            worksheet.getColumn(colIdx).width = 6;
        });

        // Column Widths
        worksheet.getColumn(1).width = 10;
        worksheet.getColumn(2).width = 25;
        worksheet.getColumn(3).width = 50;
        worksheet.getColumn(4).width = 12;
        worksheet.getColumn(5).width = 12;
        worksheet.getColumn(6).width = 12;

        // Write Tasks
        let rowIdx = 2;
        project.tasks.forEach(task => {
            const row = worksheet.getRow(rowIdx);

            // Static Data
            row.getCell(1).value = task.id; row.getCell(1).style = cellStyle;
            row.getCell(2).value = task.phase; row.getCell(2).style = phaseStyle;
            row.getCell(3).value = task.name; row.getCell(3).style = cellStyle;
            row.getCell(4).value = task.durationWeeks; row.getCell(4).style = cellStyle;
            row.getCell(5).value = task.computedStart; row.getCell(5).style = dateStyle;
            row.getCell(6).value = task.computedEnd; row.getCell(6).style = dateStyle;

            // Timeline
            if (task.computedStart && task.computedEnd) {
                const start = new Date(task.computedStart);
                const end = new Date(task.computedEnd);

                weeks.forEach((ws, i) => {
                    const colIdx = headers.length + i + 1;
                    const cell = row.getCell(colIdx);

                    const isStart = start <= ws;
                    const isEnd = end > ws;

                    if (isStart && isEnd) {
                        const endMinusOne = addDays(end, -1);
                        const nextWs = addDays(ws, 7);
                        const isConcluding = (endMinusOne >= ws && endMinusOne < nextWs);

                        if (isConcluding) {
                            if (task.type === "Milestone") { cell.value = "▲"; cell.style = markerStyle; }
                            else if (task.type === "Key Decision") { cell.value = "★"; cell.style = markerStyle; }
                            else if (task.type === "Bottleneck") { cell.value = "⚠️"; cell.style = markerStyle; }
                            else if (task.type === "External Dependency") { cell.value = "🔗"; cell.style = markerStyle; }
                            else {
                                cell.style = (task.compressionRatio <= 0.5) ? compressedBarStyle : barStyle;
                            }
                        } else {
                            if (!["Milestone", "Key Decision", "External Dependency"].includes(task.type)) {
                                cell.style = (task.compressionRatio <= 0.5) ? compressedBarStyle : barStyle;
                            } else {
                                cell.style = cellStyle;
                            }
                        }
                    } else {
                        cell.style = cellStyle;
                    }
                });
            }

            rowIdx++;
        });

        worksheet.views = [{ state: 'frozen', xSplit: 6, ySplit: 1, showGridLines: false }];

        const buffer = await workbook.xlsx.writeBuffer();
        const blob = new Blob([buffer], { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
        saveAs(blob, `${project.name.replace(/\s+/g, '_')}_Timeline.xlsx`);

    } catch (error) {
        console.error("Error generating Excel file:", error);
    }
}
